WordPress/gutenberg
# [RNMobile] Prevent `react-native-video` crash that happens when an `src` changes quickly after first being set

Fixes: https://github.com/wordpress-mobile/WordPress-iOS/issues/20882

Gutenberg Mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/6729

## What?

This PR prevents a `react-native-video`-related crash that happens when an `src` changes quickly after first being set in the Video block.

## Why?

As per https://github.com/wordpress-mobile/WordPress-iOS/issues/20882, the crash impacts iOS users. Though the crash can't be reproduced consistently, it disrupts an important flow (uploading video) for those that do encounter it. 

## How?

A video may still have the `file:` protocol for a short time after it's uploaded. In those cases [the showVideo const](https://github.com/WordPress/gutenberg/blob/fe3f484c7d7d9fadb958680ff60317b53a0bd8f7/packages/block-library/src/video/edit.native.js#L291-L294) will currently return `true` and the `src` [will be set](https://github.com/WordPress/gutenberg/blob/fe3f484c7d7d9fadb958680ff60317b53a0bd8f7/packages/block-library/src/video/edit.native.js#L325-L338). Shortly after, the `src` will be set again with the video's actual URL. It is likely this quick change to the `src` is causing the crash, as per the description of the `react-native-video` bug [here](https://github.com/wordpress-mobile/WordPress-iOS/issues/20882#issuecomment-1598228014).

With the above in mind, a check for the `file:` protocol has been added to `showVideo` in https://github.com/WordPress/gutenberg/pull/59841/commits/8cf9fd43227fba6d7f845f0a74e25446de2ec79d. Now, the `src` will not be changed quickly after it's first been set.

A [similar guard](https://github.com/wordpress-mobile/WordPress-iOS/issues/20882#issuecomment-1964763256) was added for video code in the Jetpack app, which resolved the crash in that instance and adds weight to the idea that the proposed solution here will work as intended.

The change should not have a noticeable impact on the time it takes for a video's thumbnail to be displayed, as the `file:` protocol should only be in use for a _very_ short time after the video has finished uploading (and [we do not attempt](https://github.com/WordPress/gutenberg/blob/fe3f484c7d7d9fadb958680ff60317b53a0bd8f7/packages/block-library/src/video/edit.native.js#L293) to display the thumbnail while a video's uploading). 

## Testing Instructions

The crash has been difficult to reproduce. However, we can verify that the changes don't contribute to a delay in the video's thumbnail displaying and that the `src` is not changed rapidly when first set.

1. In the WordPress iOS app, navigate to a self-hosted site without Jetpack-connected.
2. Navigate to the post editor and add a video block.
3. Upload a video to the block and verify everything loads as expected, with no change to the amount of time the thumbnail takes to display.

If testing locally, you can use the following diff to log changes to the `showVideo` and `src` variables:

```diff
diff --git a/packages/block-library/src/video/edit.native.js b/packages/block-library/src/video/edit.native.js
index f6960f0888..6bcbe246af 100644
--- a/packages/block-library/src/video/edit.native.js
+++ b/packages/block-library/src/video/edit.native.js
@@ -294,6 +294,9 @@ class VideoEdit extends Component {
 								! isUploadInProgress &&
 								! isUploadFailed;
 
+							console.log( '1. Show Video: ', showVideo );
+							console.log( '2. src: ', src );
+
 							const icon = this.getIcon(
 								isUploadFailed
 									? ICON_TYPE.RETRY
```

## Screenshots or screencast 

These side-by-side videos demonstrate that there isn't a noticeable delay in the video's thumbnail being shown. (I promise they're different videos, you can tell by the placement of the cursor 😂) 

| Before  | After |
| ------------- | ------------- |
| <video src="https://github.com/WordPress/gutenberg/assets/2998162/0cc5888c-d47f-464f-8700-254cbedd3823" /> | <video src="https://github.com/WordPress/gutenberg/assets/2998162/e75541ae-43a0-4383-8123-d7be88b60464" /> |
